kernel/sched.c

```diff
@@ -798,17 +795,27 @@ void _impl_k_sleep(s32_t duration)
 	/* wait of 0 ms is treated as a 'yield' */
 	if (duration == 0) {
 		k_yield();
-		return;
+		return 0;
 	}
 
 	ticks = _TICK_ALIGN + _ms_to_ticks(duration);
+	expected_wakeup_time = ticks + z_tick_get_32();
+
+	compiler_barrier();
 	key = irq_lock();
```

Comment:
This is unrelated to this review. If `irq_lock` (or any other kind of lock) does not introduce barriers (at least compiler barrier) then how Zerphyr's code is suppose to work?

I mentioned it once while discussing volatile and `k_sleep` in some review. There should be implicit barrier on any call to kernel function that can switch context and access hw (maybe sine other functions too). Otherwise people will be placing barriers (or worse, volatiles) in random places.

( cc @carlescufi , @ioannisg )

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/10884